### ISSUE TITLE
`jpeg_core_output_dimensions()` assumes `cinfo->master` is set, crashes when called from Skia

### ISSUE THREAD

**Issue by robn** [libjpeg-turbo/libjpeg-turbo#795]
**Have you searched the existing issues (both open and closed) in the libjpeg-turbo issue tracker to ensure that this bug report is not a duplicate?**

Yes.

**Does this bug report describe one of the [two known and unsolvable issues with the JPEG format](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf)?**

No.

**Clear and concise description of the bug:**

Note: I am a drive-by-debugger, not a direct user of libjpeg-turbo. I'm reporting a bug that I found while trying to understand why my Jellyfin server was crashing (of which I'm just a casual user). My full debugging session for that is https://github.com/jellyfin/jellyfin/issues/13016#issuecomment-2480893595. I'm not in a position to provide an isolated reproduction here though.

When compiled with `IDCT_SCALING_SUPPORTED`, `jpeg_core_output_dimensions()` assumes that `cinfo->master` is set. This appears to be different to other libjpeg implementations, and causes a crash when called from Skia, which fakes a `jpeg_decompress_struct` that does not have `master` initialised.

* `libjpeg-turbo`: https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/src/jdmaster.c#L105
* Same function in `mozjpeg`: https://github.com/mozilla/mozjpeg/blob/master/jdmaster.c#L91
* Call in Skia's `SkJpegCodec::onGetScaledDimensions`: https://skia.googlesource.com/skia/+/refs/heads/main/src/codec/SkJpegCodec.cpp#255

**Platform(s) (compiler version, operating system version, CPU) on which the bug was observed:**

FreeBSD 14.1's packaged libjpeg-turbo 3.0.4, amd64.

**Comment by dcommander** [comment]
Because the libjpeg API uses exposed structures, the only way that libjpeg-turbo could extend it in a backward compatible manner was to use the opaque master structures. Thus, we allocate those structures in `jpeg_create_compress()`/`jpeg_create_decompress()`, whereas (if I recall) libjpeg waited until `jpeg_start_compress()`/`jpeg_start_decompress()`. Any calling application that uses the API as documented will work fine, but this raises an interesting point about API hardening, which is that a calling application could do something stupid such as setting `cinfo->master` to NULL between the calls to `jpeg_create_*compress()` and `jpeg_start_*compress()`. An application could have also done something stupid with libjpeg, such as setting `cinfo->master` to NULL after calling `jpeg_start_*compress()` (which would also have probably resulted in a segfault.)

The problem is that some core libjpeg-turbo features, such as lossless compression/decompression and partial image decompression, cannot function without the master structure. So your application may work accidentally with libjpeg, but it is still abusing the API by faking the structure rather than using the API to allocate it. You can only get away with that because the API uses exposed structures. A modern C API would probably use opaque object pointers, as the TurboJPEG API does.

It is the weekend, so I am not in front of my computer. I will look into the issue in more depth tomorrow and see if there is a possible solution, but at best, it's an unsupported use case that won't ever get tested outside of your application and is thus likely to break in the future even if I fix it now. Any solution I come up with will be at the expense of some core functionality, and I can't guarantee that the library won't fail in less obvious ways if the master structure isn't available.

**Comment by dcommander** [comment]
After looking at the Skia code, it is really abusing the API in a bad way.  A calling program cannot just call an API function that accepts an object pointer without properly allocating the object and initializing its state.  Skia accidentally worked with libjpeg, but there is nothing in the definition or documentation of the libjpeg API that should have made anyone believe that it was OK to do what Skia is doing.  (Not only does Skia fail to allocate the libjpeg object, but it calls `jpeg_calc_output_dimensions()` before `jpeg_read_header()`, which is yet another API abuse.)

One of the things that `jpeg_create_compress()` and `jpeg_create_decompress()` do is initialize the memory manager, and any API function (including `jpeg_calc_output_dimensions()`) is well within its rights to use the memory manager.  Thus, any application that calls any libjpeg API function without first calling `jpeg_create_*compress()` is inviting trouble, and it is quite impossible to support such a use case in a robust manner.  As an example, a misbehaved application might set `cinfo->master` to NULL after calling `jpeg_create_*compress()`.  Such an application would accidentally work with libjpeg and segfault with libjpeg-turbo.  However, such a misbehaved application could also set `cinfo->mem` to NULL after calling `jpeg_create_*compress()`, which would cause it to segfault with both libjpeg and libjpeg-turbo.  Similarly, a misbehaved application might try to fake a libjpeg API object, such as Skia is doing, and call a global function that uses the memory manager, which would cause the application to segfault with both libjpeg and libjpeg-turbo.  Or a misbehaved application might try to fake a libjpeg API object, such as Skia is doing, and call `jpeg_calc_output_dimensions()` with an implementation of the libjpeg API that uses the libjpeg memory manager to allocate a temporary buffer in the body of `jpeg_calc_output_dimensions()`.  Hopefully that illustrates why calling applications should (a) always use `jpeg_create_*compress()` to allocate the libjpeg API object and (b) never touch opaque pointers within that object.  All of those API abuse examples are equally as bad and unsupportable as what Skia is doing.  Skia's use case isn't more supportable just because it accidentally doesn't segfault with libjpeg.

Furthermore, if Skia called `jpeg_read_header()` before `jpeg_calc_output_dimensions()`, as the API instructs, then the call to `jpeg_read_header()` would segfault both with libjpeg and libjpeg-turbo, due to the memory manager not being initialized.  Thus, Skia only works accidentally with libjpeg because it abuses the API state.  There is nothing in the definition or documentation of the libjpeg API that says `cinfo->master` will be allocated in `jpeg_start_compress()`/`jpeg_start_decompress()`.  The fact that libjpeg did so was an implementation detail.  The fact that libjpeg and libjpeg-turbo do not use the memory manager in `jpeg_calc_output_dimensions()` is also an implementation detail.  Calling programs that rely on implementation details are not, by definition, robust across different implementations.  That is the calling program's fault, not the implementation's fault.

I could modify all of the global libjpeg API functions to check whether `cinfo->master` is non-NULL before referencing it, and I could allocate the master struct in the body of `jpeg_start_*compress()` (as libjpeg does) if `cinfo->master` is NULL.  However, that would provide only the illusion of support.  It would basically create an untested and undocumented "shadow mode" of operation whose sole purpose is to support a blatant abuse of the API that, as far as I know, does not occur except in Skia.  That shadow mode of operation would not support the full feature set of libjpeg-turbo, since some of the features are attached to the master struct (which is expected to be available at any time after `jpeg_create_*compress()` is called.)  That mode would provide false confidence that applications can get away with what Skia is doing, but in the general case, that behavior is dangerous.  Again, any API function, including `jpeg_calc_output_dimensions()`, is well within its rights to use the memory manager, which is initialized in `jpeg_create_*compress()` and always has been.

libjpeg-turbo guarantees backward API/ABI compatibility with libjpeg, but that implied API contract only applies to the public documented API when used as documented.  I have bent over backwards to try and support applications that do "unique" things with the libjpeg API, even if they exceed the scope of that implied contract.  However, I have to balance the needs of one application, which is clearly abusing the API in a dangerous way, with the needs of a very ubiquitous and mission-critical JPEG library (as well as my needs as an independent open source developer who doesn't earn a salary for his work on said libary.)  In that balance, the best I can offer you in this case is to suggest either using libjpeg-turbo 2.1.x (which is still in the [Maintenance](https://libjpeg-turbo.org/DeveloperInfo/Versioning) support category but will transition to [Extended](https://libjpeg-turbo.org/DeveloperInfo/Versioning) support in January, when libjpeg-turbo 3.1 is released) or modifying the Skia code so that it does the right thing.

tl;dr: The application is attempting to treat a stateful API function as stateless.  No.  Bad.

**Comment by robn** [comment]
> tl;dr: The application is attempting to treat a stateful API function as stateless. No. Bad.

I have a lot of sympathy for this position. Thanks for looking into it! I'll go file a bug with the Skia folk. Cheers!

**Comment by make-nz** [comment]
As it was me starting
https://github.com/Thefrank/jellyfin-server-freebsd/issues/89

and I had quickly given up, I wanna say thank you for all your time!
I also get your technical explanation and general point of view of the issue(s), being perfectly reasonable and well.
Cheers

**Comment by dcommander** [comment]
Thank you.  It seems like the minimum path for Skia to work properly with libjpeg-turbo 3.x would be for it to simply call `jpeg_create_decompress()` to initialize the object before it twiddles the state, then call `jpeg_destroy_decompress()` to free the temporary object after it calls `jpeg_calc_output_dimensions()`.  It would still be abusing the API by faking the state, but it should at least work for now.  (It would still be an API abuse-- just a much lesser one than is currently being committed.)
